### PR TITLE
Library tags

### PR DESCRIPTION
**Pull request by bwlang**
This adds support for #tags in library datasets, auto-tagging on input (based on filenames) and copying tags to history datasets on import.

### CODE REVIEW

**Comment by dannon** [comment]
Nice stuff!  I'll check this out today

**Comment by bwlang** [comment]
great! - thanks @dannon.   I just committed a fix for a style violation, can you trigger the other tests (I can't easily run them locally due to tox incompatibility with conda)

**Comment by dannon** [comment]
@galaxybot test this

**Comment by dannon** [comment]
@galaxybot test this

**Comment by bwlang** [comment]
I don't think you care about what I tell you ... but I'll try anyway
@galaxybot test this

**Comment by nsoranzo** [comment]
@galaxybot test this

**Comment by nsoranzo** [comment]
@bwlang galaxybot only answers to people in the committers group.

**Comment by jmchilton** [comment]
I'll work on eliminating the transient failures. It looks like some of these are legitimate failures. @bwlang do you think you are going to have time to follow up on this? If you'd like a committer to follow up and hack on this some let us know - I can't promise anything but this is a very nice start and we should try to get this in for 17.09.

**Comment by bwlang** [comment]
@jmchilton : I have some time now... i'll at least try to suss out which tests are telling me something and which are not.

**Comment by bwlang** [comment]
@jmchilton  I think this will fix at least a few of these tests...

**Comment by bwlang** [comment]
forced a rebase to sync up with upstream/dev... hopefully will run tests now.

**Comment by nsoranzo** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by bwlang** [comment]
@galaxybot  test this ( I  think this test only works sometimes)

**Comment by bwlang** [comment]
@dannon  I think the remaining failure is actually one of the tests that is not deterministic and this is ready for review.

for the failing test... Maybe it's operating on a partial download of the tar archive?

**Comment by bwlang** [comment]
@galaxybot test this

**Comment by dannon** [comment]
@bwlang Yep, I think you're right.  I'll go through this now, thanks for the updates!

**Comment by bwlang** [comment]
@dannon , just rebased so this can apply to dev

**Comment by dannon** [comment]
Had a couple of problems with this.  sa_session wasn't defined in my testing, and I'd also like to retain tags added to data libraries from histories.  I'll try to open a PR against your branch later today.  (PR merged now, the rest of the work can be future improvements)

**Comment by bgruening** [comment]
@martenson what do you think?

**Review by martenson — APPROVED** [review]
This is a fairly narrow usecase but a solid groundwork for future improvements. Thank you Brad!

File: lib/galaxy/model/migrate/versions/0135_add_library_tags.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+# Need our custom types, but don't import anything else from model
+from galaxy.model.custom_types import TrimmedString
+
+log = logging.getLogger( __name__ )
+metadata = MetaData()
+
+
+LibraryDatasetDatasetAssociationTagAssociation_table = Table( "library_dataset_dataset_association_tag_association", metadata,
+                                                       Column("id", Integer, primary_key=True),
```
Comment (nsoranzo) [inline lib/galaxy/model/migrate/versions/0135_add_library_tags.py]: Indentation doesn't seem correct here.

**Comment by bgruening** [comment]
Double check of the migration can be found in this commit: https://github.com/bgruening/docker-galaxy-stable/commit/1eea0284ef4ec50fdebe807fa88d8898f819a39d

**Pull request merged**